lib/curvature/collector.py

```diff
@@ -0,0 +1,251 @@
+            if total < 100:
+                marker = 1
+            else:
+                marker = round(total/100)
+            sys.stderr.write("\n{} routes & ways have their coordinates added. Each '.' is 1% complete\n".format(total))
+            sys.stderr.flush()
+
+        # Add to joinable-ways.
+        for route, ways in self.routes.iteritems():
+            for way in ways:
+                way['coords'] = []
+                for ref in way['refs']:
+                    way['coords'].append(self.coords[ref])
+            # status output
+            if self.verbose:
+                i = i + 1
+                if not (i % marker):
+                    sys.stderr.write('.')
+                    sys.stderr.flush()
+
+        # Add to un-joinable ways.
+        for collection in self.collections:
+            for way in collection:
+                way['coords'] = []
+                for ref in way['refs']:
+                    way['coords'].append(self.coords[ref])
+            # status output
+            if self.verbose:
+                i = i + 1
+                if not (i % marker):
+                    sys.stderr.write('.')
+                    sys.stderr.flush()
+
+        # delete our coords database as we don't need it any more.
+        self.coords = []
+
+    # Join numbered/named routes end-to-end and add them to the way list.
+    def join_ways(self):
+        # status output
+        if self.verbose:
+            start_time = time.time()
+            i = 0
+            total = len(self.routes)
+            if total < 100:
+                marker = 1
+            else:
+                marker = round(total/100)
+            sys.stderr.write("\n{} routes will be joined, each '.' is 1% complete\n".format(total))
+            sys.stderr.flush()
+
+        for route, ways in self.routes.iteritems():
+            # status output
+            if self.verbose:
+                i = i + 1
+                if not (i % marker):
+                    sys.stderr.write('.')
+                    sys.stderr.flush()
+
+            while len(ways) > 0:
+                collection = [ways.pop()]
+                # A list of all refs added to a collection. Checking this list will
+                # prevent creating non-linar forking structures.
+                collection_refs = [collection[0]['refs']]
+                # Loop through all our ways at least as many times as we have ways
+                # to be able to catch any that join onto the end after others have
+                # been joined on.
+                j = 0
+                max_loop = len(ways)
+                # Start our first iteration with the "collection_modified" flag set to True.
+                # After this first loop, if no ways get added to the base_way,
+                # there is no reason to keep looping until max_loop
+                collection_modified = True
+                while collection_modified and j < max_loop:
```

Comment:
We should have tests for when we have `unused_ways` left over? what happens do they get discarded?